### PR TITLE
[sec_scan][27] add support for LDAP users and macOS

### PR DESCRIPTION
**Pull request by tigrato**
This PR extends support for authorized keys report for users managed by LDAP system and macOS targets.

It leverages `getpwent` to read the system database files and retrieve the user properties on macOS . On other Unix systems, it uses the reentrant`getpwent_r` although it's not (yet) standardized.

>   PLEASE NOTE: the `getpwent_r' function is not (yet) standardized.
>   The interface may change in later versions of this library.  But
>   the interface is designed following the principals used for the
>   other reentrant functions so the chances are good this is what the
>   POSIX people would choose.

Part of https://github.com/gravitational/access-graph/issues/637

### CODE REVIEW

File: lib/secretsscanner/authorizedkeys/users_list.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package authorizedkeys
+
+/*
+#cgo solaris CFLAGS: -D_POSIX_PTHREAD_SEMANTICS
```
Comment (jakule) [inline lib/secretsscanner/authorizedkeys/users_list.go:24]: solaris?

File: lib/secretsscanner/authorizedkeys/authorized_keys.go [hunk 2]
```diff
@@ -79,17 +78,17 @@ type WatcherConfig struct {
...
+	// that leverages getpwent.
+	getHostUsers func() ([]user.User, error)
 }
 
 // NewWatcher creates a new Watcher instance.
 // Returns [ErrUnsupportedPlatform] if the operating system is not supported.
 func NewWatcher(ctx context.Context, config WatcherConfig) (*Watcher, error) {
 
-	if getOS(config) != constants.LinuxOS {
+	if platform := getOS(config); platform != constants.LinuxOS && platform != constants.DarwinOS {
```
Comment (jakule) [inline lib/secretsscanner/authorizedkeys/authorized_keys.go:91]: nit: switch with empty `linux, macos` case and error as `default` is probably more readable, but it won't make a huge different.

File: lib/secretsscanner/authorizedkeys/users_list.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	"strconv"
+)
+
+// getHostUsers returns a list of all users on the host
+// from local /etc/passwd file, LDAP, or other user databases.
+func getHostUsers() (results []user.User, _ error) {
+	C.setpwent()
+	var result *C.struct_passwd
+	for {
+		result = C.getpwent()
```
Comment (jakule) [inline lib/secretsscanner/authorizedkeys/users_list.go:40]: `getpwent()` is thread safe on MacOS https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/getpwent.3.html
but it's not thread safety on Linux https://www.man7.org/linux/man-pages/man3/getpwent.3.html

Should we use a `_r` version on Linux?

Comment (tigrato) [reply lib/secretsscanner/authorizedkeys/users_list.go]: handled in [e460ace](https://github.com/gravitational/teleport/pull/45109/commits/e460ace77d36c3298e7dcd37fdb53ac1a9053691)

File: lib/secretsscanner/authorizedkeys/authorized_keys.go (hunk 2 above)
Comment (tigrato) [reply lib/secretsscanner/authorizedkeys/authorized_keys.go]: Handled in [ada9362](https://github.com/gravitational/teleport/pull/45109/commits/ada936204f13cdd26392b45126d32a65a0efb5af)

File: lib/secretsscanner/authorizedkeys/authorized_keys.go [hunk 4]
```diff
@@ -105,19 +106,16 @@ func NewWatcher(ctx context.Context, config WatcherConfig) (*Watcher, error) {
...
-		client:           config.Client,
-		logger:           config.Logger,
-		clock:            config.Clock,
-		hostID:           config.HostID,
-		usersAccountFile: config.etcPasswdFile,
+		client:       config.Client,
+		logger:       config.Logger,
+		clock:        config.Clock,
+		hostID:       config.HostID,
+		getHostUsers: config.getHostUsers,
```
Comment (zmb3) [inline lib/secretsscanner/authorizedkeys/authorized_keys.go:118]:
```suggestion
		getHostUsers: cmp.Or(config.getHostUsers, getHostUsers),
```

Then you can eliminate the nil check on line 109.

File: lib/secretsscanner/authorizedkeys/users_list_darwin.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	"os/user"
+)
+
+// getHostUsers returns a list of all users on the host
+// from local /etc/passwd file, LDAP, or other user databases.
+func getHostUsers() (results []user.User, _ error) {
+	C.setpwent()
+	var result *C.struct_passwd
+	for {
+		result = C.getpwent()
```
Comment (zmb3) [inline lib/secretsscanner/authorizedkeys/users_list_darwin.go:37]: Is it worth mentioning that `getpwent` is reentrant on macOS in a comment?

**Review by zmb3 — APPROVED** [review]

File: lib/secretsscanner/authorizedkeys/authorized_keys.go (hunk 4 above)
Comment (tigrato) [reply lib/secretsscanner/authorizedkeys/authorized_keys.go:118]: Functions headers aren't comparable

File: lib/secretsscanner/authorizedkeys/users_list_darwin.go (hunk 5 above)
Comment (tigrato) [reply lib/secretsscanner/authorizedkeys/users_list_darwin.go]: Added in [d45d8db](https://github.com/gravitational/teleport/pull/45109/commits/d45d8dbc1727088b3c31cc3f18ada7ed95bf1dd4)

**Review by jakule — APPROVED** [review]

**Pull request merged**